hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/security/client/ClientToAMTokenIdentifier.java

```diff
@@ -18,84 +18,191 @@
+        this();
+        clientToAMTokenIdentifier = new ByteArray(id, client);
+    }
+
+    public ApplicationAttemptId getApplicationAttemptID() {
+        return this.clientToAMTokenIdentifier.getApplicationAttemptID();
+    }
+
+    public String getClientName() {
+        return this.clientToAMTokenIdentifier.getClientName();
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
+        clientToAMTokenIdentifier.write(out);
     }
-    return UserGroupInformation.createRemoteUser(this.clientName.toString());
-  }
 
-  @InterfaceAudience.Private
-  public static class Renewer extends Token.TrivialRenewer {
     @Override
-    protected Text getKind() {
-      return KIND_NAME;
+    public void readFields(DataInput in) throws IOException {
+        byte[] data = IOUtils.readFullyToByteArray(in);
+        try {
+            clientToAMTokenIdentifier = new Proto();
+            clientToAMTokenIdentifier.readFields(new DataInputStream(new ByteArrayInputStream(data)));
+        } catch (InvalidProtocolBufferException e) {
+            clientToAMTokenIdentifier = new ByteArray();
+            clientToAMTokenIdentifier.readFields(new DataInputStream(new ByteArrayInputStream(data)));
+        }
+    }
+
+    @Override
+    public Text getKind() {
+        return KIND_NAME;
+    }
+
+    @Override
+    public UserGroupInformation getUser() {
+        return clientToAMTokenIdentifier.getUser();
+    }
+
+    @InterfaceAudience.Private
+    public static class Renewer extends Token.TrivialRenewer {
+        @Override
+        protected Text getKind() {
+            return KIND_NAME;
+        }
+    }
+
+    public class Proto extends ClientToAMTokenIdentifier {
+        YarnSecurityTokenProtos.ClientToAMTokenIdentifierProto proto;
+
+        public Proto() {
+        }
+
+        public Proto(ApplicationAttemptId id, String client) {
+            YarnSecurityTokenProtos.ClientToAMTokenIdentifierProto.Builder builder =
+                YarnSecurityTokenProtos.ClientToAMTokenIdentifierProto.newBuilder();
+            if (id != null) {
+                builder.setAppAttemptId(((ApplicationAttemptIdPBImpl) id).getProto());
+            }
+            if (client != null) {
+                builder.setClientName(client);
+            }
+            proto = builder.build();
+        }
+
+        public ApplicationAttemptId getApplicationAttemptID() {
+            if (!proto.hasAppAttemptId()) {
+                return null;
+            }
+            return new ApplicationAttemptIdPBImpl(proto.getAppAttemptId());
+        }
+
+        public String getClientName() {
+            return proto.getClientName();
+        }
+
+        public YarnSecurityTokenProtos.ClientToAMTokenIdentifierProto getProto() {
+            return proto;
+        }
+
+        @Override
+        public Text getKind() {
+            return KIND_NAME;
+        }
+
+        @Override
+        public UserGroupInformation getUser() {
```

Comment:
no need to override this method, the one from the parent class relies on getClientName(). overriding getClientName() in subclasses is enough